File: Common/InMotionGIT.Common/DataAcceess/CommandExecution.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;

namespace InMotionGIT.Common.DataAccess
{

    internal class CommandExecution : Interfaces.ICommandExecution
    {

        private string _commandText;
        private CommandType _commandType;
        private SqlConnection _sqlConnection;

        private string connectionName;
        private bool isLive;
        private List<SqlParameter> parameters;

        private bool useDefaultConnection;

        internal string CommandText
        {
            set
            {
                _commandText = value;
            }
        }

        internal CommandType CommandType
        {
            set
            {
                _commandType = value;
            }
        }

        internal SqlConnection SqlConnection
        {
            set
            {
                _sqlConnection = value;
            }
        }

        internal CommandExecution()
        {
            parameters = new List<SqlParameter>();
            useDefaultConnection = true;
        }

        internal CommandExecution(SqlConnection conn)
        {
            parameters = new List<SqlParameter>();
            SqlConnection = conn;
            isLive = true;
        }

        internal CommandExecution(string connName)
        {
            parameters = new List<SqlParameter>();
            useDefaultConnection = string.IsNullOrEmpty(connName);
            connectionName = connName;
        }

        public DataTable AsDataTable()
        {
            var result = new DataTable();
            ExecutionWrapper((comm) =>
            {
                var adp = new SqlDataAdapter(comm);
                adp.Fill(result);
            });
            return result;
        }

        public Hashtable AsHashTable()
        {
            Hashtable result = null;
            ExecutionWrapper((comm) => { using (IDataReader reader = comm.ExecuteReader()) { if (reader.Read()) { result = new Hashtable(reader.FieldCount); int col; var loopTo = reader.FieldCount - 1; for (col = 0; col <= loopTo; col++) { string colName = reader.GetName(col); var value = reader.GetValue(col); if (ReferenceEquals(value, DBNull.Value)) { value = null; } result.Add(colName, value); } } } });
            return result is null ? new Hashtable() : result;
        }

        public List<T> AsList<T>()
        {
            var result = new List<T>();
            ExecutionWrapper((comm) => { using (IDataReader reader = comm.ExecuteReader()) { while (reader.Read()) { var value = reader[0]; if (ReferenceEquals(value, DBNull.Value)) { value = null; } result.Add((T)value); } } });
            return result;
        }

        public List<T> AsList<T>(ModelMapper<T> mapper)
        {
            var result = new List<T>();
            ExecutionWrapper((comm) => { using (IDataReader reader = comm.ExecuteReader()) { while (reader.Read()) result.Add(mapper.Map(reader)); } });
            return result;
        }

        public void ExecuteNonQuery()
        {
            ExecutionWrapper((comm) => comm.ExecuteNonQuery());
        }

        public void ExecuteReader(Action<IDataReader> action)
        {
            ExecutionWrapper((comm) => { using (IDataReader reader = comm.ExecuteReader()) { while (reader.Read()) action.Invoke(reader); } });
        }

        public T ExecuteReaderSingle<T>(ModelMapper<T> mapper) where T : class, new()
        {
            T model = default;
            ExecutionWrapper((comm) => { using (IDataReader r = comm.ExecuteReader()) { if (r.Read()) { model = mapper.Map(r); } } });
            return model;
        }

        public void ExecuteReaderSingle(Action<IDataReader> action)
        {
            ExecutionWrapper((comm) => { using (IDataReader reader = comm.ExecuteReader()) { if (reader.Read()) { action.Invoke(reader); } } });
        }

        public T ExecuteScalar<T>()
        {
            T result = default;
            ExecutionWrapper((comm) =>
            {
                var r = comm.ExecuteScalar();
                if (ReferenceEquals(r, DBNull.Value))
                {
                    r = null;
                }
                result = (T)r;
            });
            return result;
        }

        public T ExecuteScalar<T>(T defaultValue)
        {
            T result = default;
            ExecutionWrapper((comm) =>
            {
                var r = comm.ExecuteScalar();
                if (r is null || ReferenceEquals(r, DBNull.Value))
                {
                    r = defaultValue;
                }
                result = (T)r;
            });
            return result;
        }

        private void ExecutionWrapper(Action<SqlCommand> commandAction)
        {
            var conn = GetSqlConnection();
            try
            {
                using (var comm = new SqlCommand(_commandText, conn))
                {
                    comm.CommandType = _commandType;
                    comm.CommandTimeout = Connection.DefaultTimeOut;
                    foreach (var p in parameters)
                        comm.Parameters.Add(p);
                    if (!isLive)
                    {
                        conn.Open();
                    }
                    commandAction.Invoke(comm);
                }
            }
            finally
            {
                if (!(isLive || conn is null))
                {
                    conn.Dispose();
                }
            }
        }

        private SqlConnection GetSqlConnection()
        {
            return !(_sqlConnection == null) ? _sqlConnection : useDefaultConnection ? Connection.GetDefaultConnection() : Connection.GetNamedConnection(connectionName);
        }

        public Interfaces.ICommandExecution WithParam<T>(string paramName, T paramValue)
        {
            parameters.Add(new SqlParameter(paramName, paramValue));
            return this;
        }

        public Interfaces.ICommandExecution WithParam(string paramName, object paramValue)
        {
            parameters.Add(new SqlParameter(paramName, paramValue));
            return this;
        }

        public Interfaces.ICommandExecution WithParam<T>(string paramName, SqlDbType dbType, T paramValue)
        {
            var p = new SqlParameter(paramName, dbType) { Value = paramValue };
            parameters.Add(p);
            return this;
        }

    }

}

File: Common/InMotionGIT.Common - Copy/DataAcceess/CommandExecution.vb
Imports System.Data.SqlClient

Namespace DataAccess

    Friend Class CommandExecution
        Implements Interfaces.ICommandExecution

        Private _commandText As String
        Private _commandType As CommandType
        Private _sqlConnection As SqlConnection

        Private connectionName As String
        Private isLive As Boolean
        Private parameters As List(Of SqlParameter)

        Private useDefaultConnection As Boolean

        Friend WriteOnly Property CommandText As String
            Set(ByVal value As String)
                _commandText = value
            End Set
        End Property

        Friend WriteOnly Property CommandType As CommandType
            Set(ByVal value As CommandType)
                _commandType = value
            End Set
        End Property

        Friend WriteOnly Property SqlConnection As SqlConnection
            Set(ByVal value As SqlConnection)
                _sqlConnection = value
            End Set
        End Property

        Friend Sub New()
            Me.parameters = New List(Of SqlParameter)
            Me.useDefaultConnection = True
        End Sub

        Friend Sub New(ByVal conn As SqlConnection)
            Me.parameters = New List(Of SqlParameter)
            Me.SqlConnection = conn
            Me.isLive = True
        End Sub

        Friend Sub New(ByVal connName As String)
            Me.parameters = New List(Of SqlParameter)
            Me.useDefaultConnection = String.IsNullOrEmpty(connName)
            Me.connectionName = connName
        End Sub

        Public Function AsDataTable() As DataTable Implements Interfaces.ICommandExecution.AsDataTable
            Dim result As New DataTable
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Dim adp As New SqlDataAdapter(comm)
                                    adp.Fill(result)
                                End Sub)
            Return result
        End Function

        Public Function AsHashTable() As Hashtable Implements Interfaces.ICommandExecution.AsHashTable
            Dim result As Hashtable = Nothing
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using reader As IDataReader = comm.ExecuteReader
                                        If reader.Read Then
                                            result = New Hashtable(reader.FieldCount)
                                            Dim col As Integer
                                            For col = 0 To reader.FieldCount - 1
                                                Dim colName As String = reader.GetName(col)
                                                Dim value As Object = reader.GetValue(col)
                                                If (value Is DBNull.Value) Then
                                                    value = Nothing
                                                End If
                                                result.Add(colName, value)
                                            Next col
                                        End If
                                    End Using
                                End Sub)
            Return If((result Is Nothing), New Hashtable, result)
        End Function

        Public Function AsList(Of T)() As List(Of T) Implements Interfaces.ICommandExecution.AsList
            Dim result As New List(Of T)
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using reader As IDataReader = comm.ExecuteReader
                                        Do While reader.Read
                                            Dim value As Object = reader.Item(0)
                                            If (value Is DBNull.Value) Then
                                                value = Nothing
                                            End If
                                            result.Add(DirectCast(value, T))
                                        Loop
                                    End Using
                                End Sub)
            Return result
        End Function

        Public Function AsList(Of T)(ByVal mapper As ModelMapper(Of T)) As List(Of T) Implements Interfaces.ICommandExecution.AsList
            Dim result As New List(Of T)
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using reader As IDataReader = comm.ExecuteReader
                                        Do While reader.Read
                                            result.Add(mapper.Map(reader))
                                        Loop
                                    End Using
                                End Sub)
            Return result
        End Function

        Public Sub ExecuteNonQuery() Implements Interfaces.ICommandExecution.ExecuteNonQuery
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    comm.ExecuteNonQuery()
                                End Sub)
        End Sub

        Public Sub ExecuteReader(ByVal action As Action(Of IDataReader)) Implements Interfaces.ICommandExecution.ExecuteReader
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using reader As IDataReader = comm.ExecuteReader
                                        Do While reader.Read
                                            action.Invoke(reader)
                                        Loop
                                    End Using
                                End Sub)
        End Sub

        Public Function ExecuteReaderSingle(Of T As {Class, New})(ByVal mapper As ModelMapper(Of T)) As T Implements Interfaces.ICommandExecution.ExecuteReaderSingle
            Dim model As T = CType(Nothing, T)
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using r As IDataReader = comm.ExecuteReader
                                        If r.Read Then
                                            model = mapper.Map(r)
                                        End If
                                    End Using
                                End Sub)
            Return model
        End Function

        Public Sub ExecuteReaderSingle(ByVal action As Action(Of IDataReader)) Implements Interfaces.ICommandExecution.ExecuteReaderSingle
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Using reader As IDataReader = comm.ExecuteReader
                                        If reader.Read Then
                                            action.Invoke(reader)
                                        End If
                                    End Using
                                End Sub)
        End Sub

        Public Function ExecuteScalar(Of T)() As T Implements Interfaces.ICommandExecution.ExecuteScalar
            Dim result As T = CType(Nothing, T)
            Me.ExecutionWrapper(Sub(comm As SqlCommand)
                                    Dim r As Object = comm.ExecuteScalar
                                    If (r Is DBNull.Value) Then
                                        r = Nothing
                                    End If
                                    result = DirectCast(r, T)
                                End Sub)
            Return result
        End Function

        Public Function ExecuteScalar(Of T)(ByVal defaultValue As T) As T Implements Interfaces.ICommandExecution.ExecuteScalar
            Dim result As T = CType(Nothing, T)
            ExecutionWrapper(Sub(comm As SqlCommand)
                                 Dim r As Object = comm.ExecuteScalar
                                 If ((r Is Nothing) OrElse (r Is DBNull.Value)) Then
                                     r = defaultValue
                                 End If
                                 result = DirectCast(r, T)
                             End Sub)
            Return result
        End Function

        Private Sub ExecutionWrapper(ByVal commandAction As Action(Of SqlCommand))
            Dim conn As SqlConnection = Me.GetSqlConnection
            Try
                Using comm As SqlCommand = New SqlCommand(_commandText, conn)
                    comm.CommandType = _commandType
                    comm.CommandTimeout = Connection.DefaultTimeOut
                    Dim p As SqlParameter
                    For Each p In Me.parameters
                        comm.Parameters.Add(p)
                    Next
                    If Not Me.isLive Then
                        conn.Open()
                    End If
                    commandAction.Invoke(comm)
                End Using
            Finally
                If Not (Me.isLive OrElse (conn Is Nothing)) Then
                    conn.Dispose()
                End If
            End Try
        End Sub

        Private Function GetSqlConnection() As SqlConnection
            Return If(Not IsNothing(_sqlConnection), _sqlConnection, If(Me.useDefaultConnection, Connection.GetDefaultConnection, Connection.GetNamedConnection(Me.connectionName)))
        End Function

        Public Function WithParam(Of T)(ByVal paramName As String, ByVal paramValue As T) As Interfaces.ICommandExecution
            Me.parameters.Add(New SqlParameter(paramName, paramValue))
            Return Me
        End Function

        Public Function WithParam(ByVal paramName As String, ByVal paramValue As Object) As Interfaces.ICommandExecution Implements Interfaces.ICommandExecution.WithParam
            Me.parameters.Add(New SqlParameter(paramName, paramValue))
            Return Me
        End Function

        Public Function WithParam(Of T)(ByVal paramName As String, ByVal dbType As SqlDbType, ByVal paramValue As T) As Interfaces.ICommandExecution
            Dim p As New SqlParameter(paramName, dbType) With {
                .Value = paramValue
            }
            Me.parameters.Add(p)
            Return Me
        End Function

    End Class

End Namespace